tests/test_generate_schema.py

```diff
@@ -495,15 +514,84 @@ def assert_error_messages(self, expected_error_map, error_logs):
-        # only that the number of errors matches on a per line basis.
-        # TODO: Look deeper and verify that the error message strings match as
-        # well.
         for line_number, messages in sorted(error_map.items()):
             expected_entry = expected_error_map.get(line_number)
             self.assertIsNotNone(expected_entry)
             expected_messages = expected_entry['msgs']
-            self.assertEqual(len(expected_messages), len(messages))
+            self.assertEqual(expected_messages, messages)
+
+
+class TestBigQuerySchemaToSchemaMap(unittest.TestCase):
+    def test_bq_schema_to_map_round_trip_permutations(self):
+        ''' This checks that each possible type of consititued schema, when
+            generated, then converted to a schema_map, then back to the schema,
+            they are equal.
+
+            This function is really ugly but has good coverage. This was
+            migrated from pytest fixtures which were a bit cleaner but we
+            ideally did not want to add a new dependency / library that is used
+            for testing.
+        '''
+        valid_types = BQ_TYPES
+        valid_modes = ['NULLABLE', 'REQUIRED', 'REPEATED']
+        valid_input_formats_and_modes = [('csv', True),
+                                         ('csv', False),
+                                         ('json', False)]
+        valid_keep_null_params = [True, False]
+        valid_quoted_values_are_strings = [True, False]
+        for valid_type in valid_types:
+            for valid_mode in valid_modes:
+                bq_entry = self.make_bq_schema_entry(valid_mode, valid_type)
+                schema = [bq_entry]
+                schema_map = bq_schema_to_map(schema)
+                for input_format_and_mode in valid_input_formats_and_modes:
+                    for keep_null_param in valid_keep_null_params:
+                        for quotes_are_strings in\
+                                valid_quoted_values_are_strings:
+                            generator = SchemaGenerator(
+                                input_format=input_format_and_mode[0],
+                                infer_mode=input_format_and_mode[1],
+                                keep_nulls=keep_null_param,
+                                quoted_values_are_strings=quotes_are_strings)
+                            flattened = generator.flatten_schema(schema_map)
+                            try:
+                                self.assertEqual(schema, flattened)
+                            except AssertionError as e:
+                                print("test_bq_schema_to_map_permutations"
+                                      " failed for case where: bq_entry={}\n"
+                                      "schema_generator created with values:"
+                                      "{}-{}-{}-{}"
+                                      .format(bq_entry,
```

Comment:
nit: Can you change these into f-strings? I prefer them over the old formatters, and since I now require >=3.6, I can assume that f-strings are available.

Reply:
Addressed in fae9581